Switch to use NODE_OPS_CMD for decommission and bootstrap operation

asias opened the pull request:
In commit 323f72e48a7d89d9edc40257a1f45c202accc146 (repair: Switch to
use NODE_OPS_CMD for replace operation), we switched replace operation
to use the new NODE_OPS_CMD infrastructure.

In this patch set, we continue the work to switch decommission and bootstrap 
operation to use NODE_OPS_CMD.

Fixes #8472
Fixes #8471

asias commented:
"repair: Switch to use NODE_OPS_CMD for decommission operation" is needed to wire the off-strategy for decommission operation. 

"repair: Switch to use NODE_OPS_CMD for bootstrap operation", I want this patch to be part of 4.6 which is the first release to have NODE_OPS_CMD support, so we do not need to modify the data structure send over wire. 

@avikivity I know you are going to have vacation next week. It would be nice if we can speed up the view and settle it down before your vacation. 

With "Switch to use NODE_OPS_CMD for bootstrap operation", we are more strict about avoiding unsafe concurrent topology changes, dtest still bootstrap multiple nodes in parallel, we need to change dtest a lit to make the next gate passing. With my changes to dtest + this patch set, dtest next gating works. So don't panic if a test fails to start multiple nodes in parallel.

asias commented:
On top of this series, I am going to send follow ups to

1) Wire up off-strategy for decommission
2) Share more code between different node operations for NODE_OPS_CMD

bhalevy commented:
@tgrabiec @raphaelsc please review

avikivity commented on service/storage_service.cc:
```diff
@@ -1835,15 +1836,104 @@ future<> storage_service::decommission() {
-            // FIXME: long timeout = Math.max(RING_DELAY, BatchlogManager.instance.getBatchlogTimeout());
-            auto timeout = ss.get_ring_delay();
-            ss.set_mode(mode::LEAVING, format("sleeping {} ms for batch processing and pending range setup", timeout.count()), true);
-            sleep_abortable(timeout, ss._abort_source).get();
```
What if the user did not enable repair_based_node_operations? Or if the cluster is not fully upgraded?

avikivity replied:
Ah, I see you handle non-upgraded clusters below.

asias replied:
Yes. Btw, I plan to move stream basded node operations back to removenode operations if repair_based_node_operations  is disabled. Currently, removenode only uses repair based node ops.

avikivity commented on service/storage_service.cc:
```diff
@@ -1835,15 +1836,104 @@ future<> storage_service::decommission() {
+                    auto msg = format("decommission[{}]: Nodes={} needed for decommission operation are down. It is highly recommended to fix the down nodes and try again. To proceed with best-effort mode which might cause data inconsistency, run nodetool decommission --ignore-dead-nodes <list_of_dead_nodes>. E.g., nodetool decommission --ignore-dead-nodes 127.0.0.1,127.0.0.2", uuid, nodes_down);
+                    slogger.warn("{}", msg);
+                    throw std::runtime_error(msg);
+                }
```
Shouldn't you undo the operation for the nodes that did prepare successfully?

asias replied:
We do. we throw here and  do the revert in the catch clause.

bhalevy replied:
I'd add a comment here about this assumption here.

asias replied:
We already have a comment saying we need to revert in the catch clause.

avikivity commented on service/storage_service.cc:
```diff
@@ -1835,15 +1836,104 @@ future<> storage_service::decommission() {
+                        }
+                    }
+                    slogger.debug("decommission[{}]: Stopped heartbeat_updater", uuid);
+                });
```
This (plus the code above) looks similar to the code in replace. Can it be shared?

asias replied:
I plan to do the refactor and share more code after this is merged.  I mentioned here too: https://github.com/scylladb/scylla/pull/8481#issuecomment-819268666.

avikivity commented on service/storage_service.cc:
```diff
@@ -1835,15 +1836,104 @@ future<> storage_service::decommission() {
+                // Step 6: Finish
+                req.cmd = node_ops_cmd::decommission_done;
+                parallel_for_each(nodes, [&ss, &req, uuid] (const gms::inet_address& node) {
+                    return ss._messaging.local().send_node_ops_cmd(netw::msg_addr(node), req).then([uuid, node] (node_ops_cmd_response resp) {
```
I recommend you move the inside of this lambda to a named function, so you don't have to use `ss.` everywhere.

avikivity replied:
(I mean the seastar::async call, not parallel_for_each).

asias replied:
I will consider it when I do the refactor after this series.

bhalevy commented on service/storage_service.cc:
```diff
@@ -668,6 +670,7 @@ void storage_service::bootstrap() {
                 _bootstrap_tokens, get_token_metadata_ptr(), _gossiper,
                 _sys_dist_ks.local(), get_ring_delay(), !_for_testing && !is_first_node()).get0();
 
+      if (!is_repair_based_node_ops_enabled()) {
```
please add comment about why this is done here only when RBNO is disabled.

bhalevy commented on service/storage_service.cc:
```diff
@@ -1835,15 +1840,104 @@ future<> storage_service::decommission() {
+                seastar::thread::maybe_yield();
+                if (std::find(ignore_nodes.begin(), ignore_nodes.end(), x.first) == ignore_nodes.end()) {
+                    nodes.push_back(x.first);
+                }
```
Wouldn't a `std::unordered_set<gms::inet_address> ignore_nodes` be more efficient?

bhalevy replied:
Hmm, wait, `ignore_nodes` is empty at this time.
Worth a FIXME comment as you added below in `run_bootstrap_ops`

asias replied:
To avoid big allocations if the list is large. We use list for the data we send over network. Keep it the same to avoid convert back and forth.

I added a TODO for dedecommission. We already have a TODO comment for run_bootstrap_ops.

bhalevy commented on service/storage_service.cc:
```diff
@@ -1835,15 +1840,104 @@ future<> storage_service::decommission() {
+                    throw std::runtime_error(msg);
+                }
+                if (!nodes_down.empty()) {
+                    auto msg = format("decommission[{}]: Nodes={} needed for decommission operation are down. It is highly recommended to fix the down nodes and try again. To proceed with best-effort mode which might cause data inconsistency, run nodetool decommission --ignore-dead-nodes <list_of_dead_nodes>. E.g., nodetool decommission --ignore-dead-nodes 127.0.0.1,127.0.0.2", uuid, nodes_down);
```
You can format the example in the warning message above using the actual node addresses listed in `nodes_down`.

asias replied:
The default formater will append the {} around it.  This example is clear enough in my view.

bhalevy commented on service/storage_service.cc:
```diff
@@ -1835,15 +1840,104 @@ future<> storage_service::decommission() {
+                }
+
+                // Step 3: Start heartbeat updater
+                heartbeat_updater = seastar::async([&ss, &nodes, uuid, heartbeat_updater_done] {
```
This doesn't seem like worth a thread.
Could easily be done with a `seastar::do_until` loop.

asias replied:
This is on a supper slow path. do_until makes the code harder to write. What's the gain of converting go do_until?

bhalevy replied:
The gain is with conserving resources.  The do_until state is much lighter than a full blown async thread.
and I don't think the complexity in this loop justifies a thread.

bhalevy commented on service/storage_service.cc:
```diff
@@ -1835,15 +1840,104 @@ future<> storage_service::decommission() {
+                        }).get();
+                        int nr_seconds = 10;
+                        while (!(*heartbeat_updater_done) && nr_seconds--) {
+                            sleep_abortable(std::chrono::seconds(1), ss._abort_source).get();
```
You can define another abort_source to be signaled when stopping heartbeat and just sleep_abortable the whole nr_seconds in one go, and be awakened immediately when heartbeat is stopped.

asias replied:
Actually, we can not abort here.  WE use the defer to wait for it. If we abort, we will terminate.

bhalevy commented on service/storage_service.cc:
```diff
@@ -1835,15 +1840,104 @@ future<> storage_service::decommission() {
+            std::list<gms::inet_address> ignore_nodes;
+
+            future<> heartbeat_updater = make_ready_future<>();
+            auto heartbeat_updater_done = make_lw_shared<bool>(false);
```
why do we need `heartbeat_updater_done` to be lw_shared?

asias replied:
because it is easier and safer to share between those two fibers.

bhalevy commented on service/storage_service.cc:
```diff
@@ -1835,15 +1840,104 @@ future<> storage_service::decommission() {
+                                return make_ready_future<>();
+                            });
+                        }).handle_exception([uuid] (std::exception_ptr ep) {
+                            slogger.warn("decommission[{}]: Failed to send heartbeat", uuid);
```
I think it would be more interesting to warn about this per-node, not globally.
(Presuming the error will be related to specific nodes)

asias replied:
Perhaps, I will leave it as it is for now. Since I plan to share more code as follow up.

bhalevy commented on service/storage_service.cc:
```diff
@@ -1835,15 +1840,104 @@ future<> storage_service::decommission() {
+                        // No need to revert previous prepare cmd for those who do not apply prepare cmd.
+                        return make_ready_future<>();
+                    }
+                    return ss._messaging.local().send_node_ops_cmd(netw::msg_addr(node), req).then([uuid, node] (node_ops_cmd_response resp) {
```
Is it OK to send an abort command to nodes you already succeeded sending decommission_done?
(I hope it is, given you send a uuid with it too).

If not, you can keep of list of those and ignore them too along with nodes_unknown_verb/down.

asias replied:
It is ok.

bhalevy commented on service/storage_service.cc:
```diff
@@ -1867,6 +1961,146 @@ future<> storage_service::decommission() {
+        }
+        sync_nodes.push_front(get_broadcast_address());
+
+        // Step 2: Wait until no pending node operations
```
Shouldn't we check that for all repair-based node operations?
Or is it checked somewhere else for decommission / removenode?

asias replied:
I am adding the wait logic here to make start multiple nodes in parallel to survive. We do not need it other operations.

bhalevy commented on service/storage_service.cc:
```diff
@@ -2220,6 +2463,39 @@ future<node_ops_cmd_response> storage_service::node_ops_cmd_handler(gms::inet_ad
+                        slogger.info("decommission[{}]: Added node={} as leaving node, coordinator={}", req.ops_uuid, node, coordinator);
+                        tmptr->add_leaving_endpoint(node);
+                    }
+                    return ss.update_pending_ranges(tmptr, format("decommission {}", req.leaving_nodes));
```
minor: add op uuid to the message

asias replied:
I will do it later as follow up.

bhalevy commented on service/storage_service.cc:
```diff
@@ -2220,6 +2463,39 @@ future<node_ops_cmd_response> storage_service::node_ops_cmd_handler(gms::inet_ad
+                            slogger.info("decommission[{}]: Removed node={} as leaving node, coordinator={}", req.ops_uuid, node, coordinator);
+                            tmptr->del_leaving_endpoint(node);
+                        }
+                        return ss.update_pending_ranges(tmptr, format("decommission {}", req.leaving_nodes));
```
ditto: add uuid to message

asias replied:
I will do it later as follow up.

bhalevy commented on service/storage_service.cc:
```diff
@@ -2220,6 +2463,39 @@ future<node_ops_cmd_response> storage_service::node_ops_cmd_handler(gms::inet_ad
+                        return ss.update_pending_ranges(tmptr, format("decommission {}", req.leaving_nodes));
+                    });
+                },
+                [&ss, ops_uuid] () mutable { ss.node_ops_singal_abort(ops_uuid); });
```
(preexisting) typo: s/node_ops_singal_abort/node_ops_signal_abort/

asias replied:
let's fix this as follow up.

bhalevy commented on service/storage_service.cc:
```diff
@@ -2274,13 +2550,51 @@ future<node_ops_cmd_response> storage_service::node_ops_cmd_handler(gms::inet_ad
+                        slogger.info("bootstrap[{}]: Added node={} as bootstrap, coordinator={}", req.ops_uuid, endpoint, coordinator);
+                        tmptr->add_bootstrap_tokens(tokens, endpoint);
+                    }
+                    return ss.update_pending_ranges(tmptr, format("bootstrap {}", req.bootstrap_nodes));
```
nit: add uuid to message

asias replied:
I will do it later as follow up.

bhalevy commented on service/storage_service.cc:
```diff
@@ -2274,13 +2550,51 @@ future<node_ops_cmd_response> storage_service::node_ops_cmd_handler(gms::inet_ad
+                            slogger.info("bootstrap[{}]: Removed node={} as bootstrap, coordinator={}", req.ops_uuid, endpoint, coordinator);
+                            tmptr->remove_bootstrap_tokens(tokens);
+                        }
+                        return ss.update_pending_ranges(tmptr, format("bootstrap {}", req.bootstrap_nodes));
```
nit: add uuid to message

asias replied:
I will do it later as follow up.

bhalevy reviewed (commented):
Looks good overall.  Please see individual comments for suggestions.

asias commented on repair/repair.hh:
```diff
@@ -496,6 +496,15 @@ enum class node_ops_cmd : uint32_t {
+     bootstrap_heartbeat,
+     bootstrap_abort,
+     bootstrap_done,
+     query_pending_ops,
```
I do not know if this is supposed. Since we need to modify idl files very  carefully than internal headers. Leaving it as a separate files makes more sense to me.  idl file has its own syntax, even if it looks similar to .hh file.

avikivity replied:
(the plan is to use C++ reflection to create serializers/deserializers in C++ instead of idl-compiler.py)

https://en.cppreference.com/w/cpp/keyword/reflexpr

asias commented on repair/repair.hh:
```diff
@@ -507,10 +516,29 @@ struct node_ops_cmd_request {
+            utils::UUID uuid,
+            std::list<gms::inet_address> ignore = {},
+            std::list<gms::inet_address> leaving = {},
+            std::unordered_map<gms::inet_address, gms::inet_address> replace ={},
```
fixed.

asias commented on repair/repair.hh:
```diff
@@ -507,10 +516,29 @@ struct node_ops_cmd_request {
+            std::list<gms::inet_address> ignore = {},
+            std::list<gms::inet_address> leaving = {},
+            std::unordered_map<gms::inet_address, gms::inet_address> replace ={},
+            std::unordered_map<gms::inet_address, std::list<dht::token>> bootstrap= {})
```
fixed

asias commented on service/storage_service.cc:
```diff
@@ -654,11 +654,13 @@ void storage_service::bootstrap() {
 
         // Update pending ranges now, so we correctly count ourselves as a pending replica
         // when inserting the new CDC generation.
+      if (!is_repair_based_node_ops_enabled()) {
```
done

asias commented on repair/repair.hh:
```diff
@@ -507,10 +516,29 @@ struct node_ops_cmd_request {
+        , ignore_nodes(std::move(ignore))
+        , leaving_nodes(std::move(leaving))
+        , replace_nodes(std::move(replace))
+        , bootstrap_nodes(std::move(bootstrap)) {
```
what is the real gain by adding noexcept here.

asias replied:
The problem is do we really care about the exception safeness for this class? If not, why do we bother to type noexcept everywhere. Keep simple thing simple. IMHO, it is better not to use it by default.

bhalevy replied:
what's complicated with adding noexcept where it is obvious?
similarly, why do we care about `const`? we could do without it too...

asias replied:
Sometimes maybe it is obvious and sometimes it is not. We use use a c++ feature not because it is obvious. Adding noexcept here will cause the whole scylla process to terminate if we the function actually throws.  In no ways, such user operations should cause a scylla process to terminate.

bhalevy replied:
> Adding noexcept here will cause the whole scylla process to terminate if we the function actually throws.
But it doesn't.
Maybe this case is too high level, but it's very important for lower level api's to be marked noexcept (when they are) so that their users can rely on that.  The boost library, for example, rarely uses `noexcept`, unlike the standard c++ library and it got us into trouble more than once when, for example, we require lambda functions to be nothrow move constructible and they capture a boost container which move constructor is not marked noexcept.

asias commented:
Changes in v2:
- added missing space for node_ops_cmd_request constructor  
- added comments storage_service::bootstrap for updating pending ranges and gossip status
- added todo for ignore_nodes
-  drop the use of ss._abort_source
- logged exception when it fials to send heartbeat in storage_service::decommission

avikivity commented:
I'll let @bhalevy review again.

bhalevy reviewed (approved):
I have no blocking comments.
We can fix the minor comments as a follow-up.

asias commented:
We need the dtest ccm/dtest fix: https://github.com/scylladb/scylla-ccm/pull/326 to pass the next-gating.

asias commented:
> We need the ccm/dtest fix: [scylladb/scylla-ccm#326](https://github.com/scylladb/scylla-ccm/pull/326) to pass the next-gating.

ccm/dtest fix is not needed. 

@avikivity Ping for merge.

asias commented:
@avikivity Ping for merge ^2.

avikivity commented on idl/partition_checksum.idl.hh:
```diff
@@ -129,8 +134,10 @@ struct node_ops_cmd_request {
 
 struct node_ops_cmd_response {
     bool ok;
+    std::list<utils::UUID> pending_ops;
```
Please document for which verbs this is expected to have a value.

These multiplexed verbs are hard to understand.

asias replied:
Ok. Done.

avikivity commented on idl/partition_checksum.idl.hh:
```diff
@@ -129,8 +134,10 @@ struct node_ops_cmd_request {
     std::list<gms::inet_address> leaving_nodes;
     // Map existing nodes to replacing nodes
     std::unordered_map<gms::inet_address, gms::inet_address> replace_nodes;
+    std::unordered_map<gms::inet_address, std::list<dht::token>> bootstrap_nodes;
```
So, different verbs will have different fields set in this structure?

Please add documentation.

asias replied:
Yes, documentation is added.

asias commented:
Changes in v3:
- Added more docs for node_ops_cmd_request and node_ops_cmd_response.

bhalevy commented:
@avikivity please re-review

bhalevy commented:
ping @avikivity

asias commented:
> Isn't this prone to the safety problem brought up during review for the replace operation in [#8330 (review)](https://github.com/scylladb/scylla/pull/8330#pullrequestreview-625093341) ?
> 
> For replace we concluded there was no problem if existing node reverts temporarily to the old topology on restart because the quorum in the old replica set must also be a quorum in the new replica set. During replace, in the replica set of a given partition only the node being replaced changes, and it is either permanently down (replace-with-a-different-ip) so doesn't participate in quorum, or it is the new node (replace-with-same-ip) in which case the replica set does not change.
> 
> But for bootstrap, replica set can change from (A, B, C) to (A, B, D). If some node reverts to (A, B, C), it can ACK an operation using a quorum which is not a quorum in the new replica set, e.g (B, C).
> 
> \cc @avikivity @kostja

Are you referring to the existing node restarts during bootstrap? Bootstrap does not have the issues like we have in replace. 

If an existing node restarts after all nodes ACK the bootstrap_done, it will learn the new topology from gossip during restart. All nodes will use new topology. 

If an existing node restarts before all nodes ACK the bootstrap_done, bootstrap will fail. The whole cluster will revert to use old topology.  When the existing node restarts, it will not learn the new topology. It does not matter if the node ack cql request that only satisfy the old replica set.

tgrabiec commented:
> > Isn't this prone to the safety problem brought up during review for the replace operation in [#8330 (review)](https://github.com/scylladb/scylla/pull/8330#pullrequestreview-625093341) ?
> > For replace we concluded there was no problem if existing node reverts temporarily to the old topology on restart because the quorum in the old replica set must also be a quorum in the new replica set. During replace, in the replica set of a given partition only the node being replaced changes, and it is either permanently down (replace-with-a-different-ip) so doesn't participate in quorum, or it is the new node (replace-with-same-ip) in which case the replica set does not change.
> > But for bootstrap, replica set can change from (A, B, C) to (A, B, D). If some node reverts to (A, B, C), it can ACK an operation using a quorum which is not a quorum in the new replica set, e.g (B, C).
> > \cc @avikivity @kostja
> 
> Are you referring to the existing node restarts during bootstrap? 

Yes

> Bootstrap does not have the issues like we have in replace.
> 
> If an existing node restarts after all nodes ACK the bootstrap_done, it will learn the new topology from gossip during restart. All nodes will use new topology.

But during restart we don't wait for `ring_delay`. A node could boot before the gossip change propagates to the new node. 

It's also not partition tolerant, if there is a network partition which prevents gossip to propagate, the restarted node will serve requests using old topology.

tgrabiec commented:
On Tue, May 11, 2021 at 2:29 AM Asias He ***@***.***> wrote:

> Isn't this prone to the safety problem brought up during review for the
> replace operation in #8330 (review)
> <https://github.com/scylladb/scylla/pull/8330#pullrequestreview-625093341>
> ?
> For replace we concluded there was no problem if existing node reverts
> temporarily to the old topology on restart because the quorum in the old
> replica set must also be a quorum in the new replica set. During replace,
> in the replica set of a given partition only the node being replaced
> changes, and it is either permanently down (replace-with-a-different-ip) so
> doesn't participate in quorum, or it is the new node (replace-with-same-ip)
> in which case the replica set does not change.
> But for bootstrap, replica set can change from (A, B, C) to (A, B, D). If
> some node reverts to (A, B, C), it can ACK an operation using a quorum
> which is not a quorum in the new replica set, e.g (B, C).
> \cc @avikivity <https://github.com/avikivity> @kostja
> <https://github.com/kostja>
>
> Are you referring to the existing node restarts during bootstrap?
>
> Yes
>
> Bootstrap does not have the issues like we have in replace.
> If an existing node restarts after all nodes ACK the bootstrap_done, it
> will learn the new topology from gossip during restart. All nodes will use
> new topology.
>
> But during restart we don't wait for ring_delay. A node could boot before
> the gossip change propagates to the new node.
>
> During restart, a node waits for gossip to settle.
>

It's a short wait though. Is it enough to guarantee propagation? I thought
we have much longer ring_delay wait during topology changes because
the shorter wait is not safe enough in large clusters.

It's true that ring_delay doesn't guarantee that either, but at least it's
what is considered relatively safe under normal conditions.

It's also not partition tolerant, if there is a network partition which
> prevents gossip to propagate, the restarted node will serve requests using
> old topology.
>
> Isn't this an existing issue. This can happen before this patch too. So
> far, we still use the NORMAL status.
>

But before this patch, we set the BOOTSTRAPPING state in gossip before we
start streaming. If that state doesn't reach the node, streaming will also
probably fail. After streaming, when we jump to NORMAL, if that doesn't
reach the partitioned node it will still use both old and new topology when
it learns about BOOTSTRAPPING from any neighbor, so should be safer.

If many nodes restart at the same time, the old method was also not safe
because BOOTSTRAPPING is not stored in local system tables and has to reach
us via gossip. That's https://github.com/scylladb/scylla/issues/6088

So the old method was also prone to those problems under some scenarios,
but the new method is less safe. I want us to be aware of that.

It could be made safe if we persist the state.
